adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,248 @@
+          #
+          def request_span_name
+            # uri_path_value:
+            env['PATH_INFO']
+          end
+
+          def base_url
+            if rack_request.respond_to?(:base_url)
+              rack_request.base_url
+            else
+              # Compatibility for older Rack versions
+              rack_request.url.chomp(rack_request.fullpath)
+            end
+          end
+
+          def rack_request
+            @rack_request ||= ::Rack::Request.new(env)
+          end
+
+          def record_and_reraise_error(error)
+            # TODO: implement span.set_error?
+
+            # catch exceptions that may be raised in the middleware chain
+            # Note: if a middleware catches an Exception without re raising,
+            # the Exception cannot be recorded here.
+            request_span.set_error(error) unless request_span.nil?
+            raise error
+          end
+
+          def set_attributes_after_request(span, status, headers, _response)
+            span.status = OpenTelemetry::Trace::Status.http_to_status(status)
+            span.set_attribute('http.status_code', status)
+
+            # TODO: set_attribute('http.route', ... if it's available
+            span.set_attribute('http.status_text', ::Rack::Utils::HTTP_STATUS_CODES[status])
+
+            allowed_response_headers(headers).each { |k, v| span.set_attribute(k, v) }
+          end
+
+          # @return Hash
+          def allowed_request_headers
+            {}.tap do |result|
+              Array(allowed_request_header_names).each do |header|
+                rack_header = "HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"
+                if env.key?(rack_header)
+                  result[build_attribute_name('http.request.headers.', header)] = env[rack_header]
+                end
+              end
+            end
+          end
+
+          def allowed_request_header_names
+            Array(config[:allowed_request_headers])
+          end
+
+          # @return Hash
+          def allowed_response_headers(headers)
+            return {} if headers.nil?
+
+            {}.tap do |result|
+              Array(allowed_response_header_names).each do |header|
+                new_key_name = build_attribute_name('http.response.headers.', header)
+                if headers.key?(header)
+                  result[new_key_name] = headers[header]
+                else
+                  # Try a case-insensitive lookup
+                  uppercased_header = header.to_s.upcase
+                  matching_header = headers.keys.find { |h| h.upcase == uppercased_header }
+                  if matching_header
+                    result[new_key_name] = headers[matching_header]
+                  end
+                end
+              end
+            end
+          end
+
+          def allowed_response_header_names
+            Array(config[:allowed_response_headers])
```

Comment:
Could this be memoized? Another option might be to have a more sophisticated config (rather than just a hash). We could consider trying to abstract and borrow prior art from dd-trace here.

Reply:
I've adding caching of the value in the class's singleton for the time-being.